src/shims/sync.rs

```diff
@@ -0,0 +1,438 @@
+        let writers_place = rwlock_place.offset(Size::from_bytes(8), MemPlaceMeta::None, u32_layout, this)?;
+        let readers = this.read_scalar(readers_place.into())?.to_u32()?;
+        let writers = this.read_scalar(writers_place.into())?.to_u32()?;
+        if readers != 0 || writers != 0 {
+            this.eval_libc_i32("EBUSY")
+        } else {
+            this.write_scalar(Scalar::from_u32(1), writers_place.into())?;
+            Ok(0)
+        }
+    }
+
+    fn pthread_rwlock_unlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, rwlock_op, 12)?;
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+        let rwlock_place = this.deref_operand(rwlock_op)?;
+
+        let u32_layout = this.layout_of(this.tcx.types.u32)?;
+        let readers_place = rwlock_place.offset(Size::from_bytes(4), MemPlaceMeta::None, u32_layout, this)?;
+        let writers_place = rwlock_place.offset(Size::from_bytes(8), MemPlaceMeta::None, u32_layout, this)?;
+        let readers = this.read_scalar(readers_place.into())?.to_u32()?;
+        let writers = this.read_scalar(writers_place.into())?.to_u32()?;
+        if readers != 0 {
+            this.write_scalar(Scalar::from_u32(readers - 1), readers_place.into())?;
+            Ok(0)
+        } else if writers != 0 {
+            this.write_scalar(Scalar::from_u32(0), writers_place.into())?;
+            Ok(0)
+        } else {
+            this.eval_libc_i32("EPERM")
+        }
+    }
+
+    fn pthread_rwlock_destroy(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, rwlock_op, 12)?;
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+        let rwlock_place = this.deref_operand(rwlock_op)?;
+
+        let u32_layout = this.layout_of(this.tcx.types.u32)?;
+        let readers_place = rwlock_place.offset(Size::from_bytes(4), MemPlaceMeta::None, u32_layout, this)?;
+        if this.read_scalar(readers_place.into())?.to_u32()? != 0 {
+            return this.eval_libc_i32("EBUSY");
+        }
+        let writers_place = rwlock_place.offset(Size::from_bytes(8), MemPlaceMeta::None, u32_layout, this)?;
+        if this.read_scalar(writers_place.into())?.to_u32()? != 0 {
+            return this.eval_libc_i32("EBUSY");
+        }
+
+        this.write_scalar(ScalarMaybeUndef::Undef, readers_place.into())?;
+        this.write_scalar(ScalarMaybeUndef::Undef, writers_place.into())?;
+
+        Ok(0)
+    }
+}
+
+fn check_ptr_target_min_size<'mir, 'tcx: 'mir>(ecx: &MiriEvalContext<'mir, 'tcx>, operand: OpTy<'tcx, Tag>, min_size: u64) -> InterpResult<'tcx, ()> {
```

Comment:
Since this is an assertion, please rename to `assert_*` (`check_*` sounds like it will gracefully fail).